File: com/craftinginterpreters/lox/Scanner.java
package com.craftinginterpreters.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import static com.craftinginterpreters.lox.TokenType.*;

class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0;
    private int current = 0;
    private int line = 1;

    private static final Map<String, TokenType> keywords;
    static {
        keywords = new HashMap<>();
        keywords.put("and", AND);
        keywords.put("break", BREAK);
        keywords.put("class", CLASS);
        keywords.put("else", ELSE);
        keywords.put("false", FALSE);
        keywords.put("for", FOR);
        keywords.put("fun", FUN);
        keywords.put("if", IF);
        keywords.put("inner", INNER);
        keywords.put("nil", NIL);
        keywords.put("or", OR);
        keywords.put("print", PRINT);
        keywords.put("return", RETURN);
        keywords.put("super", SUPER);
        keywords.put("this", THIS);
        keywords.put("true", TRUE);
        keywords.put("var", VAR);
        keywords.put("while", WHILE);
        keywords.put("with", WITH);
    }

    Scanner(String source) {
        this.source = source;
    }

    // Scan the source code and return a list of tokens.
    List<Token> scanTokens() {
        while (!isAtEnd()) {
            // We are at the beginning of the next lexeme.
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    // Scan a single token and add it to the list.
    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(':
                addToken(LEFT_PAREN);
                break;
            case ')':
                addToken(RIGHT_PAREN);
                break;
            case '{':
                addToken(LEFT_BRACE);
                break;
            case '}':
                addToken(RIGHT_BRACE);
                break;
            case '[':
                addToken(LEFT_BRACKET);
                break;
            case ']':
                addToken(RIGHT_BRACKET);
                break;
            case ',':
                addToken(COMMA);
                break;
            case '.':
                addToken(DOT);
                break;
            case '-':
                addToken(MINUS);
                break;
            case '+':
                addToken(PLUS);
                break;
            case ';':
                addToken(SEMICOLON);
                break;
            case '*':
                addToken(STAR);
                break;
            case '?':
                addToken(QUESTION);
                break;
            case ':':
                addToken(COLON);
                break;
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            case '/':
                if (match('/')) {
                    // A comment goes until the end of the line.
                    while (peek() != '\n' && !isAtEnd())
                        advance();
                } else if (match('*')) {
                    // C-style block comments (supporting nesting)
                    int depth = 1;
                    while (depth > 0 && !isAtEnd()) {
                        if (peek() == '/' && peekNext() == '*') {
                            advance();
                            advance();
                            depth++;
                        } else if (peek() == '*' && peekNext() == '/') {
                            advance();
                            advance();
                            depth--;
                        } else {
                            if (peek() == '\n') {
                                line++;
                            }
                            advance();
                        }
                    }
                    if (depth > 0) {
                        Lox.error(line, "Unterminated block comment.");
                    }
                } else {
                    addToken(SLASH);
                }
                break;

            case ' ':
            case '\r':
            case '\t':
                // Ignore whitespace.
                break;

            case '\n':
                line++;
                break;

            case '"':
                string();
                break;

            default:
                if (isDigit(c)) {
                    // Handle number literals.
                    number();
                } else if (isAlpha(c)) {
                    // Handle identifiers.
                    identifier();
                } else {
                    // Error handling for unexpected characters.
                    Lox.error(line, "Unexpected character.");
                }
                break;
        }
    }

    // Is the end of the source code reached?
    private boolean isAtEnd() {
        return current >= source.length();
    }

    // Consume the next character and return it.
    private char advance() {
        return source.charAt(current++);
    }

    // Advance the current position without consuming the character.
    private char peek() {
        if (isAtEnd())
            return '\0';
        return source.charAt(current);
    }

    // Peek at the next character without consuming it.
    // The scanner looks ahead at most two characters.
    private char peekNext() {
        if (current + 1 >= source.length())
            return '\0';
        return source.charAt(current + 1);
    }

    // Check if a character is a digit.
    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    // Check if a character is an alphabetic character.
    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                c == '_';
    }

    // Check if a character is alphanumeric.
    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    // Add a token to the list.
    private void addToken(TokenType type) {
        addToken(type, null);
    }

    // Add a token to the list with a literal value.
    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    // Check if the next character matches the expected character.
    private boolean match(char expected) {
        if (isAtEnd())
            return false;
        if (source.charAt(current) != expected)
            return false;
        current++;
        return true;
    }

    // Handle string literals.
    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n')
                line++;
            advance();
        }
        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.");
            return;
        }
        // The closing ".
        advance();
        // Trim the surrounding quotes.
        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    // Handle number literals.
    private void number() {
        while (isDigit(peek()))
            advance();
        // Look for a fractional part.
        if (peek() == '.' && isDigit(peekNext())) {
            // Consume the "."
            advance();
            while (isDigit(peek()))
                advance();
        }

        // Convert the string to a double.
        addToken(NUMBER,
                Double.parseDouble(source.substring(start, current)));
    }

    // Handle identifiers and reserved words.
    private void identifier() {
        while (isAlphaNumeric(peek()))
            advance();
        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null)
            type = IDENTIFIER;
        addToken(type);
    }

}

File: slox/src/main/scala/ar/edu/utn/tadp/slox/scanner/Scanner.scala
package ar.edu.utn.tadp.slox.scanner

import ar.edu.utn.tadp.slox.tokens.{Token, TokenType}
import ar.edu.utn.tadp.slox.tokens.TokenType.*
import scala.annotation.tailrec

case class Scanner(source: String) {

  private val keywords: Map[String, TokenType] = Map(
    "and" -> AND,
    "class" -> CLASS,
    "else" -> ELSE,
    "false" -> FALSE,
    "for" -> FOR,
    "fun" -> FUN,
    "if" -> IF,
    "nil" -> NIL,
    "or" -> OR,
    "print" -> PRINT,
    "return" -> RETURN,
    "super" -> SUPER,
    "this" -> THIS,
    "true" -> TRUE,
    "var" -> VAR,
    "while" -> WHILE
  )

  def scanTokens(): Either[List[String], List[Token]] = {
    scanLoop(start = 0, current = 0, line = 1, tokens = List.empty, errors = List.empty)
  }

  @tailrec
  private def scanLoop(
      start: Int,
      current: Int,
      line: Int,
      tokens: List[Token],
      errors: List[String]
  ): Either[List[String], List[Token]] = {
    if (current >= source.length) {
      val eofToken = Token(EOF, "", None, line)
      if (errors.nonEmpty) Left(errors.reverse)
      else Right((eofToken :: tokens).reverse)
    } else {
      val c = source.charAt(current)
      val nextCurrent = current + 1

      c match {
        case '(' => addToken(LEFT_PAREN, "(", None, current, nextCurrent, line, tokens, errors)
        case ')' => addToken(RIGHT_PAREN, ")", None, current, nextCurrent, line, tokens, errors)
        case '{' => addToken(LEFT_BRACE, "{", None, current, nextCurrent, line, tokens, errors)
        case '}' => addToken(RIGHT_BRACE, "}", None, current, nextCurrent, line, tokens, errors)
        case ',' => addToken(COMMA, ",", None, current, nextCurrent, line, tokens, errors)
        case '.' => addToken(DOT, ".", None, current, nextCurrent, line, tokens, errors)
        case '-' => addToken(MINUS, "-", None, current, nextCurrent, line, tokens, errors)
        case '+' => addToken(PLUS, "+", None, current, nextCurrent, line, tokens, errors)
        case ';' => addToken(SEMICOLON, ";", None, current, nextCurrent, line, tokens, errors)
        case '*' => addToken(STAR, "*", None, current, nextCurrent, line, tokens, errors)

        case '!' =>
          if (matchChar('=', nextCurrent)) addToken(BANG_EQUAL, "!=", None, current, nextCurrent + 1, line, tokens, errors)
          else addToken(BANG, "!", None, current, nextCurrent, line, tokens, errors)

        case '=' =>
          if (matchChar('=', nextCurrent)) addToken(EQUAL_EQUAL, "==", None, current, nextCurrent + 1, line, tokens, errors)
          else addToken(EQUAL, "=", None, current, nextCurrent, line, tokens, errors)

        case '<' =>
          if (matchChar('=', nextCurrent)) addToken(LESS_EQUAL, "<=", None, current, nextCurrent + 1, line, tokens, errors)
          else addToken(LESS, "<", None, current, nextCurrent, line, tokens, errors)

        case '>' =>
          if (matchChar('=', nextCurrent)) addToken(GREATER_EQUAL, ">=", None, current, nextCurrent + 1, line, tokens, errors)
          else addToken(GREATER, ">", None, current, nextCurrent, line, tokens, errors)

        case '/' =>
          if (matchChar('/', nextCurrent)) {
            val endLinePos = skipComment(nextCurrent + 1)
            scanLoop(endLinePos, endLinePos, line, tokens, errors)
          } else {
            addToken(SLASH, "/", None, current, nextCurrent, line, tokens, errors)
          }

        case ' ' | '\r' | '\t' =>
          scanLoop(nextCurrent, nextCurrent, line, tokens, errors)

        case '\n' =>
          scanLoop(nextCurrent, nextCurrent, line + 1, tokens, errors)

        case '"' =>
          scanString(nextCurrent, line, tokens, errors)

        case digit if digit.isDigit =>
          scanNumber(current, line, tokens, errors)

        case alpha if alpha.isLetter || alpha == '_' =>
          scanIdentifier(current, line, tokens, errors)

        case _ =>
          val err = s"[line $line] Error: Unexpected character '$c'."
          scanLoop(nextCurrent, nextCurrent, line, tokens, err :: errors)
      }
    }
  }

  private def matchChar(expected: Char, current: Int): Boolean = {
    if (current >= source.length) false
    else source.charAt(current) == expected
  }

  private def skipComment(current: Int): Int = {
    var p = current
    while (p < source.length && source.charAt(p) != '\n') {
      p += 1
    }
    p
  }

  private def addToken(
      tType: TokenType,
      lexeme: String,
      literal: Option[Any],
      start: Int,
      nextCurrent: Int,
      line: Int,
      tokens: List[Token],
      errors: List[String]
  ): Either[List[String], List[Token]] = {
    val token = Token(tType, lexeme, literal, line)
    scanLoop(nextCurrent, nextCurrent, line, token :: tokens, errors)
  }

  private def scanString(
      current: Int,
      line: Int,
      tokens: List[Token],
      errors: List[String]
  ): Either[List[String], List[Token]] = {
    var p = current
    var currentLine = line
    while (p < source.length && source.charAt(p) != '"') {
      if (source.charAt(p) == '\n') currentLine += 1
      p += 1
    }

    if (p >= source.length) {
      val err = s"[line $currentLine] Error: Unterminated string."
      scanLoop(p, p, currentLine, tokens, err :: errors)
    } else {
      val valStr = source.substring(current, p)
      val fullLexeme = source.substring(current - 1, p + 1)
      val token = Token(STRING, fullLexeme, Some(valStr), currentLine)
      val nextPos = p + 1
      scanLoop(nextPos, nextPos, currentLine, token :: tokens, errors)
    }
  }

  private def scanNumber(
      start: Int,
      line: Int,
      tokens: List[Token],
      errors: List[String]
  ): Either[List[String], List[Token]] = {
    var p = start
    while (p < source.length && source.charAt(p).isDigit) {
      p += 1
    }

    if (p < source.length && source.charAt(p) == '.' && p + 1 < source.length && source.charAt(p + 1).isDigit) {
      p += 1 // consume '.'
      while (p < source.length && source.charAt(p).isDigit) {
        p += 1
      }
    }

    val lexeme = source.substring(start, p)
    val numVal = lexeme.toDouble
    val token = Token(NUMBER, lexeme, Some(numVal), line)
    scanLoop(p, p, line, token :: tokens, errors)
  }

  private def scanIdentifier(
      start: Int,
      line: Int,
      tokens: List[Token],
      errors: List[String]
  ): Either[List[String], List[Token]] = {
    var p = start
    while (p < source.length && (source.charAt(p).isLetterOrDigit || source.charAt(p) == '_')) {
      p += 1
    }

    val text = source.substring(start, p)
    val tType = keywords.getOrElse(text, IDENTIFIER)
    val token = Token(tType, text, None, line)
    scanLoop(p, p, line, token :: tokens, errors)
  }
}
